build/generate-licenses.go

```diff
@@ -97,26 +102,97 @@ func main() {
 			continue
 		}
 
-		if strings.HasPrefix(filepath.Base(hdr.Name), "README") {
+		fileBaseName := filepath.Base(hdr.Name)
+		licenseName := strings.TrimSuffix(fileBaseName, ".txt")
+
+		if strings.HasPrefix(fileBaseName, "README") {
 			continue
 		}
 
-		if strings.HasPrefix(filepath.Base(hdr.Name), "deprecated_") {
+		if strings.HasPrefix(fileBaseName, "deprecated_") {
 			continue
 		}
-		out, err := os.Create(path.Join(destination, strings.TrimSuffix(filepath.Base(hdr.Name), ".txt")))
+		out, err := os.Create(path.Join(destination, licenseName))
 		if err != nil {
 			log.Fatalf("Failed to create new file. %s", err)
 		}
 
 		defer out.Close()
 
-		if _, err := io.Copy(out, tr); err != nil {
+		// some license files have same content, so we need to detect these files and create a convert map into a json file
+		// Later we use this convert map to avoid adding same license content with different license name
+		h := md5.New()
+		// calculate md5 and write file in the same time
+		r := io.TeeReader(tr, h)
+		if _, err := io.Copy(out, r); err != nil {
 			log.Fatalf("Failed to write new file. %s", err)
 		} else {
 			fmt.Printf("Written %s\n", out.Name())
+
+			md5 := hex.EncodeToString(h.Sum(nil))
+			aliasesFiles[md5] = append(aliasesFiles[md5], licenseName)
+		}
+	}
+
+	// generate convert license name map
+	licenseAliases := make(map[string]string)
+	for _, fileNames := range aliasesFiles {
+		if len(fileNames) > 1 {
+			licenseName := getLicenseNameFromAliases(fileNames)
+			for _, fileName := range fileNames {
+				licenseAliases[fileName] = licenseName
+			}
 		}
 	}
+	// save convert license name map to file
+	b, err := json.Marshal(licenseAliases)
+	if err != nil {
+		log.Fatalf("Failed to create json bytes. %s", err)
+		return
+	}
+	f, err := os.Create(licenseAliasesDestination)
+	if err != nil {
+		log.Fatalf("Failed to create license aliases json file. %s", err)
+	}
+	defer f.Close()
+	if _, err = f.Write(b); err != nil {
+		log.Fatalf("Failed to write license aliases json file. %s", err)
+	}
 
 	fmt.Println("Done")
 }
+
+func getLicenseNameFromAliases(fnl []string) string {
```

Comment:
It looks like a strange algorithm, if it is designed as so intentionally, it needs enough tests to show how it works.

Reply:
Yes, it is. But I didn't find a good solution to solve it.
The root reason is #20915, where is the first time reported and I found it today.
Maybe we only need to calculate MD5. as this function will only be executed during the building.

Reply:
I didn't say it is not acceptable, I mean "it needs enough tests to show how it works."

Reply:
OK, I will try to add some tests.

Reply:
4e76dbff4d1aa14602de8710bbd59876851af2fd

We can not 100% guarantee that all licenses with same content have same prefix.
But it make sense that they should.
I added a check in this commit: if they don't have same prefix, it will return error.
Then we can ensure this prerequisite works unless it returns error.